File: apps/reading/src/api/writeTwinCollectiveSettingsDraftFullscreenNdMoCompose.ts
/**
 * Write-mode twin collective analysis + settings draft fullscreen weekly ND (pure).
 *
 * Operator vision: after floating multi-agent research on the draft+fullscreen
 * pack, merge twin substrate + completed chases into written analysis while
 * settings add-model inventory remains available — never writes analysis/draft,
 * never mutates inventory, never live-routes.
 *
 * draft_written / analysis_written / merge_executed always false.
 * secrets_stored / inventory_mutated always false.
 * live_dispatched / pack_dispatched / backlog_mutated / store_mutated always false.
 * production_router_verdict always REJECT; live_router_authorized always false.
 */

import {
  composeWriteModeTwinCollectiveAnalysis,
  type WriteModeTwinCollectiveAnalysisCompose,
  type WriteModeTwinCollectiveAnalysisInput,
} from "./writeModeTwinCollectiveAnalysisCompose";
import {
  composeSettingsAddModelDraftFullscreenWeeklyNdMo,
  type SettingsAddModelDraftFullscreenWeeklyNdMoCompose,
  type SettingsAddModelDraftFullscreenWeeklyNdMoInput,
} from "./settingsAddModelDraftFullscreenWeeklyNdMoCompose";

export interface WriteTwinCollectiveSettingsDraftFullscreenNdMoInput {
  write: Omit<WriteModeTwinCollectiveAnalysisInput, "operator_ack">;
  settings_research: Omit<
    SettingsAddModelDraftFullscreenWeeklyNdMoInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose {
  session_id: string;
  parent_asset_id: string;
  week_id: string;
  write: WriteModeTwinCollectiveAnalysisCompose;
  settings_research: SettingsAddModelDraftFullscreenWeeklyNdMoCompose;
  pack_ready: boolean;
  draft_written: false;
  analysis_written: false;
  merge_executed: false;
  secrets_stored: false;
  inventory_mutated: false;
  live_dispatched: false;
  pack_dispatched: false;
  backlog_mutated: false;
  store_mutated: false;
  production_router_verdict: "REJECT";
  live_router_authorized: false;
  purchase_executed: false;
  twin_written: false;
  live_execution_authorized: false;
  notes: string[];
  authority: "write_twin_collective_settings_draft_fullscreen_nd_mo_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose write twin collective analysis with settings draft fullscreen weekly ND.
 * Never writes analysis/draft; never mutates settings inventory; ND REJECT.
 */
export function composeWriteTwinCollectiveSettingsDraftFullscreenNdMo(
  input: WriteTwinCollectiveSettingsDraftFullscreenNdMoInput,
): WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.write || typeof input.write !== "object") {
    throw new Error("write must be an object");
  }
  if (!input.settings_research || typeof input.settings_research !== "object") {
    throw new Error("settings_research must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "draft_written=false · analysis_written=false · merge_executed=false",
    "secrets_stored=false · inventory_mutated=false",
    "live_dispatched=false · pack_dispatched=false · backlog_mutated=false · store_mutated=false",
    "production_router_verdict=REJECT · live_router_authorized=false",
  ];

  const write = composeWriteModeTwinCollectiveAnalysis({
    ...input.write,
    operator_ack: input.operator_ack,
  });
  notes.push(...write.notes.map((n) => `[write] ${n}`));

  const settings_research = composeSettingsAddModelDraftFullscreenWeeklyNdMo({
    ...input.settings_research,
    operator_ack: input.operator_ack,
  });
  notes.push(
    ...settings_research.notes.map((n) => `[settings_research] ${n}`),
  );

  const session_id = requireNonEmpty(write.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    write.parent_asset_id,
    "parent_asset_id",
  );
  const week_id = requireNonEmpty(settings_research.week_id, "week_id");

  const aligned =
    settings_research.session_id === session_id &&
    settings_research.parent_asset_id === parent_asset_id;
  if (!aligned) {
    notes.push(
      "session/parent mismatch between write and settings_research — pack_ready blocked",
    );
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      aligned &&
      write.pack_ready === true &&
      settings_research.pack_ready === true &&
      settings_research.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      aligned &&
      input.operator_ack === true &&
      settings_research.production_router_verdict === "REJECT" &&
      (write.pack_ready === true || settings_research.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — write twin collective + settings draft fullscreen ND ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — write, settings_research, alignment, or operator_ack gate open",
    );
  }

  if (
    write.draft_written !== false ||
    write.analysis_written !== false ||
    write.merge_executed !== false ||
    write.store_mutated !== false ||
    write.live_dispatched !== false ||
    settings_research.secrets_stored !== false ||
    settings_research.inventory_mutated !== false ||
    settings_research.draft_written !== false ||
    settings_research.merge_executed !== false ||
    settings_research.live_dispatched !== false ||
    settings_research.production_router_verdict !== "REJECT" ||
    settings_research.live_router_authorized !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("merge_executed=false");
  notes.push("secrets_stored=false");
  notes.push("inventory_mutated=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("production_router_verdict=REJECT");
  notes.push("live_router_authorized=false");
  notes.push("purchase_executed=false");
  notes.push("twin_written=false");
  notes.push("live_execution_authorized=false");

  return {
    session_id,
    parent_asset_id,
    week_id,
    write,
    settings_research,
    pack_ready,
    draft_written: false,
    analysis_written: false,
    merge_executed: false,
    secrets_stored: false,
    inventory_mutated: false,
    live_dispatched: false,
    pack_dispatched: false,
    backlog_mutated: false,
    store_mutated: false,
    production_router_verdict: "REJECT",
    live_router_authorized: false,
    purchase_executed: false,
    twin_written: false,
    live_execution_authorized: false,
    notes,
    authority:
      "write_twin_collective_settings_draft_fullscreen_nd_mo_compose_advisory",
  };
}

export function formatWriteTwinCollectiveSettingsDraftFullscreenNdMoSummary(
  c: WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `write_ready=${c.write.pack_ready} · ` +
    `settings_research_ready=${c.settings_research.pack_ready} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `analysis_written=false · draft_written=false · merge_executed=false`
  );
}

File: substrate/write_twin_collective_settings_draft_fullscreen_nd_mo_compose.py
"""Write twin collective + settings draft fullscreen weekly ND pack (pure).

draft_written / analysis_written / merge_executed always False.
secrets_stored / inventory_mutated always False.
live_dispatched / pack_dispatched / backlog_mutated / store_mutated always False.
production_router_verdict always REJECT; live_router_authorized always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.settings_add_model_draft_fullscreen_weekly_nd_mo_compose import (
    SettingsAddModelDraftFullscreenWeeklyNdMoCompose,
    SettingsAddModelDraftFullscreenWeeklyNdMoComposeError,
    compose_settings_add_model_draft_fullscreen_weekly_nd_mo,
)
from substrate.write_mode_twin_collective_analysis_compose import (
    WriteModeTwinCollectiveAnalysisCompose,
    WriteModeTwinCollectiveAnalysisComposeError,
    compose_write_mode_twin_collective_analysis,
)


class WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(ValueError):
    """Fail-closed validation for write twin + settings draft fullscreen ND."""


@dataclass(frozen=True)
class WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose:
    session_id: str
    parent_asset_id: str
    week_id: str
    write: WriteModeTwinCollectiveAnalysisCompose
    settings_research: SettingsAddModelDraftFullscreenWeeklyNdMoCompose
    pack_ready: bool
    draft_written: bool
    analysis_written: bool
    merge_executed: bool
    secrets_stored: bool
    inventory_mutated: bool
    live_dispatched: bool
    pack_dispatched: bool
    backlog_mutated: bool
    store_mutated: bool
    production_router_verdict: str
    live_router_authorized: bool
    purchase_executed: bool
    twin_written: bool
    live_execution_authorized: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "week_id": self.week_id,
            "write": self.write.to_dict(),
            "settings_research": self.settings_research.to_dict(),
            "pack_ready": self.pack_ready,
            "draft_written": False,
            "analysis_written": False,
            "merge_executed": False,
            "secrets_stored": False,
            "inventory_mutated": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "backlog_mutated": False,
            "store_mutated": False,
            "production_router_verdict": "REJECT",
            "live_router_authorized": False,
            "purchase_executed": False,
            "twin_written": False,
            "live_execution_authorized": False,
            "notes": list(self.notes),
            "authority": (
                "write_twin_collective_settings_draft_fullscreen_nd_mo_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_write_twin_collective_settings_draft_fullscreen_nd_mo(
    *,
    write: object,
    settings_research: object,
    operator_ack: object,
    require_both: object | None = None,
) -> WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose:
    """Write twin collective + settings draft fullscreen ND. Never writes."""
    if not isinstance(operator_ack, bool):
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(write, dict):
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            "write must be an object"
        )
    if not isinstance(settings_research, dict):
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            "settings_research must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "draft_written=false · analysis_written=false · merge_executed=false",
        "secrets_stored=false · inventory_mutated=false",
        "live_dispatched=false · pack_dispatched=false · backlog_mutated=false · store_mutated=false",
        "production_router_verdict=REJECT · live_router_authorized=false",
    ]

    try:
        w = compose_write_mode_twin_collective_analysis(
            session_id=write.get("session_id"),
            draft_id=write.get("draft_id"),
            parent_asset_id=write.get("parent_asset_id"),
            twin_slices=write.get("twin_slices"),
            chase_slots=write.get("chase_slots"),
            analysis_kind=write.get("analysis_kind"),
            operator_ack=operator_ack,
            base_draft_html=write.get("base_draft_html"),
            extra_findings=write.get("extra_findings"),
            require_both=write.get("require_both"),
        )
    except WriteModeTwinCollectiveAnalysisComposeError as e:
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            str(e)
        ) from e
    notes.extend(f"[write] {n}" for n in w.notes)

    try:
        sr = compose_settings_add_model_draft_fullscreen_weekly_nd_mo(
            settings=settings_research.get("settings"),
            research_pack=settings_research.get("research_pack"),
            operator_ack=operator_ack,
            require_both=settings_research.get("require_both"),
        )
    except SettingsAddModelDraftFullscreenWeeklyNdMoComposeError as e:
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            str(e)
        ) from e
    notes.extend(f"[settings_research] {n}" for n in sr.notes)

    session = _require_nonempty(w.session_id, field="session_id")
    parent = _require_nonempty(w.parent_asset_id, field="parent_asset_id")
    week = _require_nonempty(sr.week_id, field="week_id")

    aligned = sr.session_id == session and sr.parent_asset_id == parent
    if not aligned:
        notes.append(
            "session/parent mismatch between write and settings_research — pack_ready blocked"
        )

    if require:
        pack_ready = (
            aligned
            and w.pack_ready is True
            and sr.pack_ready is True
            and sr.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            aligned
            and operator_ack is True
            and sr.production_router_verdict == "REJECT"
            and (w.pack_ready is True or sr.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — write twin collective + settings draft fullscreen ND ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — write, settings_research, alignment, or operator_ack gate open"
        )

    if (
        w.draft_written is not False
        or w.analysis_written is not False
        or w.merge_executed is not False
        or w.store_mutated is not False
        or w.live_dispatched is not False
        or sr.secrets_stored is not False
        or sr.inventory_mutated is not False
        or sr.draft_written is not False
        or sr.merge_executed is not False
        or sr.live_dispatched is not False
        or sr.production_router_verdict != "REJECT"
        or sr.live_router_authorized is not False
    ):
        raise WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "draft_written=false",
            "analysis_written=false",
            "merge_executed=false",
            "secrets_stored=false",
            "inventory_mutated=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "backlog_mutated=false",
            "store_mutated=false",
            "production_router_verdict=REJECT",
            "live_router_authorized=false",
            "purchase_executed=false",
            "twin_written=false",
            "live_execution_authorized=false",
        )
    )

    return WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose(
        session_id=session,
        parent_asset_id=parent,
        week_id=week,
        write=w,
        settings_research=sr,
        pack_ready=pack_ready,
        draft_written=False,
        analysis_written=False,
        merge_executed=False,
        secrets_stored=False,
        inventory_mutated=False,
        live_dispatched=False,
        pack_dispatched=False,
        backlog_mutated=False,
        store_mutated=False,
        production_router_verdict="REJECT",
        live_router_authorized=False,
        purchase_executed=False,
        twin_written=False,
        live_execution_authorized=False,
        notes=tuple(notes),
        authority=(
            "write_twin_collective_settings_draft_fullscreen_nd_mo_compose_advisory"
        ),
    )


def format_write_twin_collective_settings_draft_fullscreen_nd_mo_summary(
    c: WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"write_ready={c.write.pack_ready} · "
        f"settings_research_ready={c.settings_research.pack_ready} · "
        f"verdict={c.production_router_verdict} · "
        f"analysis_written=false · draft_written=false · merge_executed=false"
    )


__all__ = [
    "WriteTwinCollectiveSettingsDraftFullscreenNdMoCompose",
    "WriteTwinCollectiveSettingsDraftFullscreenNdMoComposeError",
    "compose_write_twin_collective_settings_draft_fullscreen_nd_mo",
    "format_write_twin_collective_settings_draft_fullscreen_nd_mo_summary",
]
